system/lib/wasmfs/backends/ignore_case_backend.cpp

```diff
@@ -20,105 +37,171 @@ std::string normalize(const std::string& name) {
   }
   return result;
 }
-} // namespace
+
+} // anonymous namespace
 
 namespace wasmfs {
 
-// Problem: Child entries are stored both in IgnoreCaseDirectory and in baseDirectory.
-//
-// Original name case preservation could be possible if MemoryDirectory::entries
-// were accesible. Then, we would store the original case there but search
-// ignoring case.
-//
-class IgnoreCaseDirectory : public MemoryDirectory {
-  using BaseClass = MemoryDirectory;
-  std::shared_ptr<Directory> baseDirectory;
+class IgnoreCaseDirectory : public VirtualDirectory {
 
-public:
-  IgnoreCaseDirectory(std::shared_ptr<Directory> base, backend_t proxyBackend)
-    : BaseClass(base->locked().getMode(), proxyBackend), baseDirectory(base) {}
+  struct ChildInfo {
+    std::string originalName;
+    std::shared_ptr<File> child;
+  };
 
-  std::shared_ptr<File> getChild(const std::string& name) override {
-    return BaseClass::getChild(normalize(name));
-  }
+  // Map normalized names to virtual files and their non-normalized names.
+  std::map<std::string, ChildInfo> children;
 
+public:
+  IgnoreCaseDirectory(std::shared_ptr<Directory> real, backend_t backend)
+    : VirtualDirectory(real, backend) {}
+
+  std::shared_ptr<File> getChild(const std::string& name) override;
   std::shared_ptr<DataFile> insertDataFile(const std::string& name,
-                                           mode_t mode) override {
-    auto name2 = normalize(name);
-    auto baseDirLocked = baseDirectory->locked();
-    auto child = baseDirLocked.insertDataFile(name2, mode);
-    if (child) {
-      insertChild(name2, child);
-      // Directory::Hanlde needs a parent
-      child->locked().setParent(cast<Directory>());
```

Comment:
Why don't we set a parent anymore?

Reply:
Great catch. I thought the backend-independent code would set the parent, but that was wrong. I fixed this bug and added a test. I also noticed that we weren't properly clearing the parents of overwritten directories, so I fixed that and added another test.